Fixes and suggestions for DefrostedTuna: https://github.com/Defrosted…

javajon opened the pull request:
Hi Rick!,

What you have done in kubernetes-scaffolding is brilliant and demonstrates your command of Linux and Kubernetes. Very impressive.

While experimenting with the script I stumbled across a few bumps and smoothed them out for your consideration. Here are my notes:

- Update the Jenkins plugin versions:
-- Credentials Binding to 1.18	
-- Git to 3.9.3 (security fix)
-- Kubernetes to 1.15.2
-- Pipeline: Job to 2.32

- Per new Jenkins chart requirements changed jenkins-values.yaml to lower camel case propertyKeys (not upper camel case/pascal case)
- Removed `rbac.install: true` section from jenkins-value.yaml as it now defaults to true and `install` changed to `create`. Also part of the Jenkins chart requirements.
- Cert manager install failed due to new missing prerequisites. install-cert-manager.sh needed these added prerequisites as defined [here](https://github.com/helm/charts/tree/master/stable/cert-manager).
- In install-cert-manager.sh changed to kubectl --watch for deployment instead of bash with loop.  Not critical, just an example of another way to do it with less "bash-ing".
- Jenkins ingress was not being added
- Add to readme the zsh on MacOS is needed as the scripts use both `brew` and `print`.
- Added to readme a note ad end on how to remove the provisioned cluster.
- Fixed echo to use parameters in init-cluster.sh
- Formatting of script welcome text.

------------

Future considerations:

- Put this in a Docker container and avoid zsh, print, MacOS/brew dependencies.
- Output all the important information like passwords, keys ,and URLs to a results file.
- Perhaps change to Ansible instead of bash
- Publish an article on the DigitalOcean community board

Great stuff,

Jonathan

javajon committed:
Fixes and suggestions for DefrostedTuna: https://github.com/DefrostedTuna

javajon committed:
Added install report

javajon committed:
Remove harbor-helm dir

javajon committed:
Mispelling and removed mistaken comment about zsh.

javajon committed:
Experimenting with ExternalDNS

Commit:
Fixed Helm by adding update. Added spinner, more details in readme notes.

Commit:
Enhanced machine dependencies. Improved installation prompts.

javajon committed:
Updated documentation.

javajon committed:
Update readme.md

Commit:
Added wait times.

Commit:
Polishing for TLS with Jenkins and Harbor.

Commit:
Added regcred secret key for pushing from Jenkins to Harbor registry.

Commit:
Fixed error at end of jenkins script.

Commit:
Drafting Spinnaker. Fixed spinner logic. Some more polish.

Commit:
Added an ignore

Commit:
Made admin password generation length longer

DefrostedTuna commented:
Hey Jonathan!

I apologize for it taking so incredibly long to get back to you about this. I ended up having to put my projects on hold shortly after uploading this and I haven't had much of a chance to sit down and work on anything until recently.

My original intentions for this script was to simply make it a bit easier to spin up a cluster. It's great to see someone else take an interest in the project.

The changes in this look absolutely phenomenal. I see you caught quite a bit of my typos and fixed a couple of things I had forgotten to remove/revert after testing. The additions to the script look great as well! Wrapping the project into a Docker container and moving things over to Ansible are things I had considered while building the project, but I hadn't quite gotten that far just yet lol. I've only gotten the chance to skim through the code at the moment, however I should have the time to clone it and tinker around with everything sometime over the weekend.

I'd love to collaborate on this project and extended it even further than it already has been. It was pretty fun getting the script up and running initially. There's a ton that could be done to streamline something like this. I imagine it would be a pretty handy tool for those who work heavily with Kubernetes on DigitalOcean.

javajon commented:
Hi Rick,

Cool, you are real ;-) Thanks for getting back to me. First, your script
was an excellent foundation I needed at just the right time.  I was in the
midst of writing something like this and you greatly help me to bring in my
schedule.  Thank you, thank you.  I learned so much from your approach.

I sent you that pull request, but since then also polished it up some
more.  Take a gander at
https://github.com/javajon/kubernetes-pipeline-on-doks.  I have been using
this script to generate and rip down about a dozen clusters,  I spent a
bunch of time on Ingress with TLS, all works now. Your script saved me so
much time and I have peace of mind knowing that if I ever need to create a
new DOKS cluster this script will have my back.

I wanted to understand Ansible some more so this week I have a playbook
that provisions Harbor using its REST API. Playbooks are a good follow up
for things that could not or should not be done with Helm and Kubernetes
configuration.  Yes, I can imagine calling the DO, Helm and Kube rest API
calls using Ansible instead of tieing to OS dependent Bash with the CLI
tools.

Great idea to add Harbor, I really like it,  I have used Sonatype and
Artifactory before, Harbor has what I need: Containers, Charts and *Clair*
with just the right amount of account management and container promotion
workflow.

I'm planning to add playbooks for Jenkins. I'm not a big fan of Jenkins as
it's a sucking black hole of time and energy for engineers, Jenkins-X looks
nice, but only if you like to ride their train that has limited station
stops. I sort of like Drone for CI because of its sheer simplicity, but
more excited about what people are going to do with Knative and Tekton.

I am planning to add Spinnaker for CD and an optional Istio Mesh, mainly to
provide monitoring. This would come fairly close to completing most peoples
needs for CI/CD for K8s on DO.  There is also Terraform which could make
this less DO centric.  There is also a notion of a CI/CD Kubernetes
Operator that could set all this up, declaratively. However, there is much
to be said for just simple Bash and get-er-done. Well done sir.

Yes, this could be taken further. Let's see where it goes. Stay in touch,

- Jonathan Johnson
https://www.linkedin.com/in/javajon/



On Thu, May 23, 2019 at 5:45 PM Rick Bennett <notifications@github.com>
wrote:

> Hey Jonathan!
>
> I apologize for it taking so incredibly long to get back to you about
> this. I ended up having to put my projects on hold shortly after uploading
> this and I haven't had much of a chance to sit down and work on anything
> until recently.
>
> My original intentions for this script was to simply make it a bit easier
> to spin up a cluster. It's great to see someone else take an interest in
> the project.
>
> The changes in this look absolutely phenomenal. I see you caught quite a
> bit of my typos and fixed a couple of things I had forgotten to
> remove/revert after testing. The additions to the script look great as
> well! Wrapping the project into a Docker container and moving things over
> to Ansible are things I had considered while building the project, but I
> hadn't quite gotten that far just yet lol. I've only gotten the chance to
> skim through the code at the moment, however I should have the time to
> clone it and tinker around with everything sometime over the weekend.
>
> I'd love to collaborate on this project and extended it even further than
> it already has been. It was pretty fun getting the script up and running
> initially. There's a ton that could be done to streamline something like
> this. I imagine it would be a pretty handy tool for those who work heavily
> with Kubernetes on DigitalOcean.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/DefrostedTuna/kubernetes-scaffolding/pull/1?email_source=notifications&email_token=ACIEDPHBTKKKCVW2FCHWE3TPW4GADA5CNFSM4HIEUX7KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODWDSNFY#issuecomment-495396503>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACIEDPCSUHSAKLCH2IXUDPDPW4GADANCNFSM4HIEUX7A>
> .
>


-- 

   - - Jonathan
   Jonathan JohnsonJava Specialist, Software EngineeringDijure, LLC
   <http://www.linkedin.com/in/javajon/>

Commit:
Bumped versions for: Jenkins k8s plugin and Harbor chart.

Commit:
Added questions for Harbor volume sizes. Drafting addition of Vault.

Commit:
Bumped Jenkins plugin versions

DefrostedTuna commented:
Hey Jonathan, I finally got the free time to go through and try out everything you've done here. I was able to spin up a couple of clusters over the weekend and tinker around with how everything has been tweaked. I'm definitely pleased with all of the changes and additions. 

The core functionality is still there with the bonus of having things cleaned up and streamlined a bit more. I like the addition of Vault and Spinnaker as well. I've used Vault previously with my employer and am pleased with it, although I've never used it in a Kubernetes hosted environment. As for Spinnaker, this is the first I'm hearing of that one. I did a bit of research on it though and it seems like a pretty robust platform for managing deployments and piplelines. It reminds me of a more streamlined version of [Rancher](https://rancher.com/).

With that said, I did run into a couple of issues with getting the script to set everything up properly. Namely with the Vault and Spinnaker installations. Also with some other minor things as well though.

When creating a cluster from scratch and allowing the creation of volumes (Jenkins and Harbor), I got tossed a couple of errors in regards to the creation of the volumes. They did however get created properly despite the errors. Following suit with the volumes, I think it would be a bonus to allow the volumes sizes to be customizable when they are created via the script. I believe it was set up this way originally, but it was fairly cumbersome to choose all of the sizes. I'm namely thinking of the volume for the Docker Registry in Harbor. Having a default size of 250gb is good for hosting lots of images, but for a hobbyist it seems like a bit of overkill.

Vault didn't toss me any errors with the script, however the container wouldn't move to a ready state out of the box. After the script finished, the Vault container would hang on `CrashLoopBackoff`. I'm not sure if it's an issue with the configuration and how it's being set up in the script, or maybe it's an issue with the container itself. I didn't spend a significant amount of time looking into that one as I opted to use most of my time exploring Spinnaker since it was a shiny new toy I've yet to see lol.

Spinnaker gave me the most trouble. It looks as though you're still tinkering around with getting it set up, so I assume that's why things went awry with the installation. Halyard installed without an issue, but everything after that just killed the script for me. I'm not familiar with Spinnaker, so I spent the majority of my weekend looking into what can be done with it and how to set it up. I followed the guide that you had linked in the code comments in an attempt to get everything set up without much luck. Took a look at the official documentation as well and checked out a few other guides without much success either. Seems as though a lot of people are having issues with getting Spinnaker up and running unfortunately. It's a shame because it seems like such a promising platform. I feel as thought this would be able to replace Jenkins for a good amount of situations. I'm sure most people can agree that it would be a good thing to be able to get rid of Jenkins

I'll keep looking into resolving some of the issues when I get some more free time.

javajon commented:
Hi Rick,

Me too, my time is stretched.

Yes, my goal is to maintain the existing features and simply add more
options and features.  Like adding things such as Vault and Spinnaker.
Yes, those two are still in draft and I'm putting up a new version today
that notes these as draft versions.  The Vault one does hang. Just not
baked yet (time ;-)). Today I'm adding a new service that replicates
secrets to other namespaces. e.g. replicate the "regcred" to Jenkins
namespace so it can talk to Harbor.

I have an intern working with me this summer and he is writing Ansible
provisioning scripts that are internal for us that occur after this generic
DOKS provisioning.  For instance, we have Ansible scripts that work with
the Harbor REST API to provision things like users, registries, robot
accounts and Slack messaging.

Now that I understand Ansible better, I see this project evolving, with the
same functionality of course.

The errors provisioning the volumes could be due to DO, as it's not the
most stable API/service. I have not noticed them, but let keep our eyes
peeled.

The volume sizes should still be prompted. hm..   read -e -p "For Harbor
volume ${VOLUME_NAME}, specify the desired size, [${VOLUME_SIZE}]GB: " -i
"${VOLUME_SIZE}" VOLUME_SIZE

Yes, the volume sizes are important to ask.  It makes sense to keep the
volume size small for the first timers, but DO does not yet allow you to
increase existing volume sizes.  I asked DO, they said not yet. Within 2
weeks we ran out of room in the registry.  I figured drive space is
relatively not that costly and if you are a first time, most likely the
cattle cluster would shortly be slaughtered.

I'm planning on adding to the project

   - K8s Configmap/secrets replicator -
   https://github.com/mittwald/kubernetes-replicator   (although I would
   like to create something similar using K8s Operator pattern that is more
   secure)
   - Istio Mesh chart with Prometheus, Alert Manager, and Grafana
   - Metrics-server helm chart:
   https://github.com/helm/charts/tree/master/stable/metrics-server
   - I try to keep versions up to date, such as Jenkins and its silly
   pluggin hell. For instance, the Jenkins K8s plugin was just updated to
   1.16.0.  Side note, I'm not a big fan of Jenkins because it's old and
   bloated, however, I have high hopes for https://cd.foundation/ so better
   things could be written based on Tekton.  Also, like ideas from Argos and
   Drone for CI.

I'm mulling the idea of adding the OpenLDAP via Helm chart so there is a
fundamental user list with permissions since both Harbor and Jenkins could
benefit with a central user directory.

Your choice to use Harbor was great. It's a really good project.

Spinnaker is not really meant for full CI/CD, just CD.  I look at Jenkins,
Argos, Drone, Draft, Trellis, CircleCI as best in the CI camp. They build
artifacts and get the containers into the registry.  Ideas like Spinnaker
in conjunction with Helm and Istio are great CD solutions.  I see Harbor as
the keystone between CI / CD.  Some like to have full tools like Jenkins-X
take care of it all, but it feels unflexible like an express train that
does not stop at local, custom stops.

Jonathan

On Mon, Jun 10, 2019 at 2:48 PM Rick Bennett <notifications@github.com>
wrote:

> Hey Jonathan, I finally got the free time to go through and try out
> everything you've done here. I was able to spin up a couple of clusters
> over the weekend and tinker around with how everything has been tweaked.
> I'm definitely pleased with all of the changes and additions.
>
> The core functionality is still there with the bonus of having things
> cleaned up and streamlined a bit more. I like the addition of Vault and
> Spinnaker as well. I've used Vault previously with my employer and am
> pleased with it, although I've never used it in a Kubernetes hosted
> environment. As for Spinnaker, this is the first I'm hearing of that one. I
> did a bit of research on it though and it seems like a pretty robust
> platform for managing deployments and piplelines. It reminds me of a more
> streamlined version of Rancher <https://rancher.com/>.
>
> With that said, I did run into a couple of issues with getting the script
> to set everything up properly. Namely with the Vault and Spinnaker
> installations. Also with some other minor things as well though.
>
> When creating a cluster from scratch and allowing the creation of volumes
> (Jenkins and Harbor), I got tossed a couple of errors in regards to the
> creation of the volumes. They did however get created properly despite the
> errors. Following suit with the volumes, I think it would be a bonus to
> allow the volumes sizes to be customizable when they are created via the
> script. I believe it was set up this way originally, but it was fairly
> cumbersome to choose all of the sizes. I'm namely thinking of the volume
> for the Docker Registry in Harbor. Having a default size of 250gb is good
> for hosting lots of images, but for a hobbyist it seems like a bit of
> overkill.
>
> Vault didn't toss me any errors with the script, however the container
> wouldn't move to a ready state out of the box. After the script finished,
> the Vault container would hang on CrashLoopBackoff. I'm not sure if it's
> an issue with the configuration and how it's being set up in the script, or
> maybe it's an issue with the container itself. I didn't spend a significant
> amount of time looking into that one as I opted to use most of my time
> exploring Spinnaker since it was a shiny new toy I've yet to see lol.
>
> Spinnaker gave me the most trouble. It looks as though you're still
> tinkering around with getting it set up, so I assume that's why things went
> awry with the installation. Halyard installed without an issue, but
> everything after that just killed the script for me. I'm not familiar with
> Spinnaker, so I spent the majority of my weekend looking into what can be
> done with it and how to set it up. I followed the guide that you had linked
> in the code comments in an attempt to get everything set up without much
> luck. Took a look at the official documentation as well and checked out a
> few other guides without much success either. Seems as though a lot of
> people are having issues with getting Spinnaker up and running
> unfortunately. It's a shame because it seems like such a promising
> platform. I feel as thought this would be able to replace Jenkins for a
> good amount of situations. I'm sure most people can agree that it would be
> a good thing to be able to get rid of Jenkins
>
> I'll keep looking into resolving some of the issues when I get some more
> free time.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/DefrostedTuna/kubernetes-scaffolding/pull/1?email_source=notifications&email_token=ACIEDPDJZU4BJVDT4NXELZLPZ2OXFA5CNFSM4HIEUX7KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODXK3C2Q#issuecomment-500543850>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACIEDPHRIMEEFCW35TCDSETPZ2OXFANCNFSM4HIEUX7A>
> .
>


-- 

   - - Jonathan Johnson
   Software Architectures ★ Presenter ★ Kubernetes & Java Specialist ★ Dijure,
   LLC <http://www.linkedin.com/in/javajon/>

DefrostedTuna commented:
Yeah I was assuming that Vault and Spinnaker were drafts based on the comments in the code for the time being. I really like the idea of having Vault for secret management, but maybe I'm biased because I'm familiar with the platform and the workflows that come along with developing around it haha. Spinnaker isn't a platform I've personally used, but from what I've read, it seems would be a phenomenal tool to manage deployments. I'm really stoked to tinker with that one. 

Ansible is still a mystery to me at the moment. A lot of people reference Ansible and Terraform together, so I need to get around to digging into both of those platforms at some point. I'm aware that Terraform is used for creating network infrastructure through code, but that's about the extent of my knowledge regarding it. Kubernetes and CI/CD are actually all fairly new to me. I've been casually working with them and learning everything in my free time since ~Q4 of 2018. 

With that said I'm still trying to work out a good CI/CD workflow that makes sense and is flexible enough to adapt to a variety of application types. I've used Jenkins to map a pretty robust deployment strategy to AWS before, but I'd like to stay away from Jenkins this time if I can. I really enjoy the idea of having a modular approach to the tools available when it comes to workflows as opposed to a single platform that handles everything (looking at you Jenkins). That approach seems to give more flexibility to what can be accomplished, and doesn't restrict the design to only what a single platform chooses to support.

In regards to the issues I ran into with the script; I did track down the volume error and found that it was due to the syntax not being supported in the native version of Bash that ships with MacOS. Swapped the syntax around and it worked as intended. Haven't tested it on my Linux machine just yet though. With the volume size limitation in DigitalOcean at the moment, I do think that having a default of ~250GB is perfectly fine. Whenever I'm spinning up clusters to try out something new, I usually toss it down to just a few GB. Keeping a good size main repository is ideal though. Now that I figured out what was going on with the volume size prompt I'm not too worried about all of that.

OpenLDAP would be a pretty good choice for integrating users I'd say. I've only worked with things on a small scale at the moment, but I can see things easily getting out of hand with a large influx of users. As for Harbor, I'm pretty pleased with it. I spent a good amount of time deciding on a private Docker repository to drop in. All of the open source products I ran across didn't have the feature set I was looking for in a scalable solution. Harbor crammed everything I was looking for into one package.

One thing I would like to note in regards to Harbor is that Docker doesn't like pulling images from it if you're using the staging Let's Encrypt certificate. Since it's self signed, Docker throws a fit and won't let you grab anything even if you've authenticated with the system (x509). I found that out the hard way a few days ago. Spent way too much time trying to figure out a clean way to handle that. Spoiler: it wasn't a clean fix when it comes to Kubernetes unfortunately. That would probably be a good thing to note in the readme at some point.

Ideally, as a barebones setup goes, I'd like to have some sort of CI platform, a private docker repository, and a good CD platform as part of a main pipeline. Bonuses would be supporting tools like Vault and other things to manage and visualize the cluster as a whole. I'm not sure of an open source CI platform that could be self hosted at the moment. The big contenders here that I've seen a lot of are TravisCI, CircleCI and GitlabCI. I've personally used TravisCI and don't have a problem continuing to use it in all honesty. Harbor has the repository covered, and Spinnaker seems like it would be a good solution for the CD platform.

I should have a bit more free time this weekend to work on the script. I'll toss a PR or two up if I get to a good stopping point so we can get some of the fixes in.

Commit:
Preserve Harbor org secret in kube-system. Added k8s secret replicator. Added text saying Vault and Spinnaker are experimental. Updated Jenkins plugin version.

Commit:
Adjusted logic for Harbor robot secret 'regcred' with secret replicator

DefrostedTuna committed:
Read syntax support for Bash 3.x

DefrostedTuna committed:
Update Harbor install method

javajon committed:
Merge pull request #1 from DefrostedTuna/harbor-updates

Harbor updates

Commit:
Added replicator annotation for harbor admin secret to harbor namespace.

Commit:
Bumped Kubernetes plugin for Jenkins to 1.16.1

Commit:
added a script to install the sonarqube helm chart on k8s and a yaml file that acts as a template
with various configurations to be made to sonarqube upon installation including: plugins and service type

Commit:
Revamped the install-sonarqube script to handle both staging and production, it also creates the necessary secrets/configmaps in K8s to store sonarqube data, added a stage to the init script to prompt the user to install sonarqube on K8s, fixed a small bugged in install-jenkins where -f flag was not added, modified the sonarqube template values to have more up to date plugin versions

Commit:
added some minor language change to one of the descriptive comments in the script

javajon committed:
Merge pull request #2 from williamEstony/sonarqube

added a script to install the sonarqube helm chart on k8s and a yaml …

Commit:
Tightened security for Jenkins install
